types/json8-patch/index.d.ts

```diff
@@ -59,23 +60,215 @@ export type JsonPatchOperation = AddOperation | RemoveOperation | ReplaceOperati
  */
 export type JsonPatch = JsonPatchOperation[];
 
-export interface ApplyResult {
-    doc: any;
+export interface ApplyResult<T> {
+    doc: T;
```

Comment:
I can see what you're trying to capture here, but I think that type doesn't work for functions such as `apply`: the document returned doesn't need to have the same type/shape as the one in the input. For instance:
```ts
let doc1 = { a: 1 }; // doc1: { a: number }
let doc2 = apply(doc1, [{ "path": "/a", "op": "replace", "value": 26 }]).doc;
```
In that example, `doc2` looks like `{ a: 26 }`, and no longer has the *same type* as `doc1`, so we can't say that both `doc1` and `doc2` have some type `T`. We can't know the type of the returned doc, because it depends on the patch/operations applied.

So I think instead of using generics, the way to go is to use `unknown` where we had `any` before.

Reply:
I see your point. I was fixated on my specific use case I missed that one.

For this particular interface it doesn't rally mean much to have a generic applied here. The magic would be in the apply function. What if I would change it to:

```typescipt
export function apply<S=unknown, T=unknown>(doc: S, patch: JsonPatch, options?: { reversible: false }): ApplyResult<V>;
```

The reason being, now I can have a variable declared somewhere like `const apply = oo.apply<SrcType, TrgType>;` so I have my types defined without casting the type each time I use the apply function. The default behavior would be what you suggesting which is the `unknown` so you still can cast the type when calling the apply function.

Reply:
I understand. I think in general we'd like to avoid having type parameters that don't appear in the parameter types, because as they can't be inferred from the argument types, it's essentially a cast. And it's a cast that might confuse or go unnoticed by other people, so we lean towards avoiding it. So I think I'd prefer if you don't use generics.
And you could still have something like `const apply = oo.apply as (doc: SrcType, ...) => TrgType;`, I think.

Reply:
Understood. Thanks for clarifying. I'll make an update in the morning.

Reply:
I've changed the generics to `any` and `unknown` where applicable.